File: custom_adapter/TopOnAdapter/Objc/6.5.x/Interstitial/AlxTopOnInterstitialAdapter.m
//
//  AlxTopOnInterstitialAdapter.m
//  AlxAdsOCDemo
//

#import "AlxTopOnInterstitialAdapter.h"
#import "AlxTopOnBaseManager.h"
#import "AlxTopOnInterstitialDelegate.h"
#import "AlxTopOnInterstitialEvent.h"
#import "AlxTopOnBiddingRequest.h"
#import "AlxTopOnBiddingRequestManager.h"
#import "AlxTopOnTool.h"

static NSString *const TAG = @"AlxTopOnInterstitialAdapter";

@interface AlxTopOnInterstitialAdapter ()
@property (nonatomic, strong) AlxInterstitialAd *interstitialAd;
@property (nonatomic, strong) AlxTopOnInterstitialDelegate *interstitialDelegate;
@end

@implementation AlxTopOnInterstitialAdapter

#pragma mark - lazy load
- (AlxTopOnInterstitialDelegate *)interstitialDelegate {
    if (_interstitialDelegate == nil) {
        _interstitialDelegate = [[AlxTopOnInterstitialDelegate alloc] init];
        _interstitialDelegate.adStatusBridge = self.adStatusBridge;
    }
    return _interstitialDelegate;
}

#pragma mark - Ad load
- (void)loadADWithArgument:(ATAdMediationArgument *)argument {
    NSLog(@"%@: loadADWithArgument", TAG);
    NSLog(@"%@: loadAD: isMainThread=%@", TAG, [NSThread isMainThread] ? @"YES" : @"NO");
    
    NSString *bidId = argument.serverContentDic[kATAdapterCustomInfoBuyeruIdKey];
    NSLog(@"%@: loadAD: bidId=%@", TAG, bidId ?: @"空");
    
    dispatch_async(dispatch_get_main_queue(), ^{
        NSString *unitKeyStr = [AlxTopOnBaseManager unitID];
        NSString *unitId = argument.serverContentDic[unitKeyStr];
        
        if (!unitId || unitId.length == 0) {
            NSString *errorStr = @"unitid is empty";
            NSLog(@"%@: loadAD: error = %@", TAG, errorStr);
            NSError *error = [AlxTopOnBaseManager errorWithCode:-100 message:errorStr];
            [self.adStatusBridge atOnAdLoadFailed:error adExtra:@{}];
            return;
        }
        NSLog(@"%@: loadAD: unitid = %@", TAG, unitId);
        
        if (bidId) {
            // Bidding 场景：从缓存中取出已加载的广告 / Bidding scenario: retrieve the pre-loaded ad from cache
            AlxTopOnBiddingRequest *biddingRequest = [[AlxTopOnTool shared] getRequestItemWithUnitID:unitId];
            if (biddingRequest) {
                self.interstitialAd = (AlxInterstitialAd *)biddingRequest.customObject;
                
                if (self.interstitialAd) {
                    NSLog(@"%@: loadAD: bid ad loaded, notify success", TAG);
                    // ⚠️ 注意：将广告对象传给 TopOn SDK / Note: pass the ad object to TopOn SDK
                    NSMutableDictionary *adExtra = [NSMutableDictionary dictionary];
                    adExtra[kATAdAssetsCustomObjectKey] = self.interstitialAd;
                    [self.adStatusBridge atOnInterstitialAdLoadedExtra:adExtra];
                } else {
                    NSLog(@"%@: loadAD: bid ad object is empty", TAG);
                    NSError *error = [NSError errorWithDomain:@"AlxTopOnAdapter" code:-100 userInfo:@{NSLocalizedDescriptionKey: @"Bid ad object is empty"}];
                    [self.adStatusBridge atOnAdLoadFailed:error adExtra:@{}];
                }
            } else {
                NSLog(@"%@: loadAD: bid request not found in cache", TAG);
                NSError *error = [NSError errorWithDomain:@"AlxTopOnAdapter" code:-100 userInfo:@{NSLocalizedDescriptionKey: @"Bid request not found"}];
                [self.adStatusBridge atOnAdLoadFailed:error adExtra:@{}];
            }
            [[AlxTopOnTool shared] removeRequestItemWithUnitID:unitId];
        } else {
            // 普通加载场景 / Normal loading scenario
            self.interstitialAd = [[AlxInterstitialAd alloc] init];
            self.interstitialAd.delegate = self.interstitialDelegate;
            // 设置 Delegate 的广告对象引用 / Set the delegate's ad object reference
            self.interstitialDelegate.interstitialAd = self.interstitialAd;
            
            NSLog(@"%@: start loading ad with unitId: %@", TAG, unitId);
            [self.interstitialAd loadAdWithAdUnitId:unitId];
        }
    });
}

#pragma mark - C2S Bidding
+ (void)bidRequestWithPlacementModel:(ATPlacementModel *)placementModel
                      unitGroupModel:(ATUnitGroupModel *)unitGroupModel
                                info:(NSDictionary *)info
                          completion:(void (^)(ATBidInfo * _Nullable, NSError * _Nullable))completion {
    NSLog(@"%@: bidRequestWithPlacementModel", TAG);
    
    dispatch_async(dispatch_get_main_queue(), ^{
        if (![AlxTopOnBaseManager isInitialized]) {
            [AlxTopOnBaseManager initSDKWithServerInfo:info];
        }
    });
    
    AlxTopOnInterstitialEvent *customEvent = [[AlxTopOnInterstitialEvent alloc] initWithInfo:info localInfo:info];
    customEvent.isC2SBiding = YES;
    
    AlxTopOnBiddingRequest *request = [[AlxTopOnBiddingRequest alloc]
                                       initWithUnitGroup:unitGroupModel
                                       customEvent:customEvent
                                       unitID:info[[AlxTopOnBaseManager unitID]]
                                       placementID:placementModel.placementID
                                       extraInfo:info
                                       adType:ATAdFormatInterstitial
                                       bidCompletion:completion];
    
    [[AlxTopOnBiddingRequestManager shared] startWithRequest:request];
}

#pragma mark - Ad Ready Check (实例方法 / Instance Method)
- (BOOL)adReadyInterstitialWithInfo:(NSDictionary *)info {
    NSLog(@"%@: adReadyInterstitialWithInfo", TAG);
    
    // 检查广告对象是否准备好 / Check if the ad object is ready
    if (self.interstitialAd) {
        // 可以调用 isReady 或其他检查方法 / Can call isReady or other check methods
        NSLog(@"%@: adReady = YES", TAG);
        return YES;
    }
    
    NSLog(@"%@: adReady = NO", TAG);
    return NO;
}

#pragma mark - Show Ad (实例方法 / Instance Method)
- (void)showInterstitialInViewController:(UIViewController *)viewController {
    NSLog(@"%@: showInterstitialInViewController", TAG);
    
    dispatch_async(dispatch_get_main_queue(), ^{
        if (self.interstitialAd) {
            // 展示广告 / Show the ad
            [self.interstitialAd showAdWithPresent:viewController];
            NSLog(@"%@: ad shown", TAG);
        } else {
            NSLog(@"%@: interstitialAd is nil", TAG);
        }
    });
}

@end

File: custom_adapter/TopOnAdapter/Swift/6.5.x/Interstitial/AlxTopOnInterstitialAdapter.swift
//
//  AlxTopOnInterstitialAdapter.swift
//

import Foundation
import AnyThinkSDK
import AlxAds

@objc(AlxTopOnInterstitialAdapter)
public class AlxTopOnInterstitialAdapter: AlxTopOnBaseAdapter, ATBaseInterstitialAdapterProtocol {
    
    private static let TAG = "AlxTopOnInterstitialAdapter"
    
    private var interstitialAd: AlxInterstitialAd?
    private lazy var interstitialDelegate: AlxTopOnInterstitialDelegate = {
        let delegate = AlxTopOnInterstitialDelegate()
        delegate.adStatusBridge = self.adStatusBridge
        return delegate
    }()
    
    // MARK: - ATBaseInterstitialAdapterProtocol
    
    @objc public var adStatusBridge: ATInterstitialAdStatusBridge!
    
    // MARK: - Ad Load
    
    @objc public override func loadAD(with argument: ATAdMediationArgument) {
        NSLog("%@: loadAD(with:)", AlxTopOnInterstitialAdapter.TAG)
        NSLog("%@: loadAD: isMainThread=%@", AlxTopOnInterstitialAdapter.TAG, Thread.current.isMainThread ? "YES" : "NO")
        
        let bidId = argument.serverContentDic[kATAdapterCustomInfoBuyeruIdKey] as? String
        NSLog("%@: loadAD: bidId=%@", AlxTopOnInterstitialAdapter.TAG, bidId ?? "空")
        
        DispatchQueue.main.async {
            let unitKeyStr = AlxTopOnBaseManager.unitID
            guard let unitId = argument.serverContentDic[unitKeyStr] as? String, !unitId.isEmpty else {
                let errorStr = "unitid is empty"
                NSLog("%@: loadAD: error = %@", AlxTopOnInterstitialAdapter.TAG, errorStr)
                let error = AlxTopOnBaseManager.error(code: -100, msg: errorStr)
                self.notifyLoadFailed(error: error)
                return
            }
            NSLog("%@: loadAD: unitid = %@", AlxTopOnInterstitialAdapter.TAG, unitId)
            
            if let bidId = bidId {
                // Bidding 场景：从缓存中取出已加载的广告
                if let biddingRequest = AlxTopOnTool.shared.getRequestItem(withUnitID: unitId) as? AlxTopOnBiddingRequest {
                    self.interstitialAd = biddingRequest.customObject as? AlxInterstitialAd
                    
                    if let interstitialAd = self.interstitialAd {
                        NSLog("%@: loadAD: bid ad loaded, notify success", AlxTopOnInterstitialAdapter.TAG)
                        var adExtra: [AnyHashable: Any] = [:]
                        adExtra[kATAdAssetsCustomObjectKey] = interstitialAd
                        self.notifyInterstitialLoaded(adExtra: adExtra)
                    } else {
                        NSLog("%@: loadAD: bid ad object is empty", AlxTopOnInterstitialAdapter.TAG)
                        let error = NSError(domain: "AlxTopOnAdapter", code: -100, userInfo: [NSLocalizedDescriptionKey: "Bid ad object is empty"])
                        self.notifyLoadFailed(error: error)
                    }
                } else {
                    NSLog("%@: loadAD: bid request not found in cache", AlxTopOnInterstitialAdapter.TAG)
                    let error = NSError(domain: "AlxTopOnAdapter", code: -100, userInfo: [NSLocalizedDescriptionKey: "Bid request not found"])
                    self.notifyLoadFailed(error: error)
                }
                AlxTopOnTool.shared.removeRequestItem(withUnitID: unitId)
            } else {
                // 普通加载场景
                self.interstitialAd = AlxInterstitialAd()
                self.interstitialAd?.delegate = self.getInterstitialDelegate()
                self.interstitialDelegate.interstitialAd = self.interstitialAd
                
                NSLog("%@: start loading ad with unitId: %@", AlxTopOnInterstitialAdapter.TAG, unitId)
                self.interstitialAd?.loadAd(adUnitId: unitId)
            }
        }
    }
    
    private func getInterstitialDelegate() -> AlxTopOnInterstitialDelegate {
        self.interstitialDelegate.adStatusBridge = self.adStatusBridge
        return self.interstitialDelegate
    }
    
    // MARK: - Dynamic Invocation Helper Methods
    
    private func notifyInterstitialLoaded(adExtra: [AnyHashable: Any]) {
        if let bridge = self.adStatusBridge {
            let selector = NSSelectorFromString("atOnInterstitialAdLoadedExtra:")
            if bridge.responds(to: selector) {
                bridge.perform(selector, with: adExtra)
            }
        }
    }
    
    private func notifyLoadFailed(error: Error) {
        if let bridge = self.adStatusBridge {
            let selector = NSSelectorFromString("atOnAdLoadFailed:adExtra:")
            if bridge.responds(to: selector) {
                let emptyDict: [AnyHashable: Any] = [:]
                bridge.perform(selector, with: error, with: emptyDict)
            }
        }
    }
    
    // MARK: - C2S Bidding
    
    @objc public static func bidRequestWithPlacementModel(_ placementModel: ATPlacementModel,
                                                          unitGroupModel: ATUnitGroupModel,
                                                          info: [AnyHashable: Any],
                                                          completion: @escaping (ATBidInfo?, Error?) -> Void) {
        NSLog("%@: bidRequestWithPlacementModel", AlxTopOnInterstitialAdapter.TAG)
        
        DispatchQueue.main.async {
            if !AlxTopOnBaseManager.isInitialized {
                AlxTopOnBaseManager.initSDK(serverInfo: info)
            }
        }
        
        let customEvent = AlxTopOnInterstitialEvent(info: info, localInfo: info)
        customEvent.isC2SBiding = true
        
        let request = AlxTopOnBiddingRequest(
            unitGroup: unitGroupModel,
            customEvent: customEvent,
            unitID: info[AlxTopOnBaseManager.unitID] as? String,
            placementID: placementModel.placementID,
            extraInfo: info,
            adType: ATAdFormat.interstitial,
            bidCompletion: completion
        )
        
        AlxTopOnBiddingRequestManager.shared.start(with: request)
    }
    
    // MARK: - Ad Ready Check (实例方法)
    
    @objc public func adReadyInterstitial(withInfo info: [AnyHashable: Any]) -> Bool {
        NSLog("%@: adReadyInterstitialWithInfo", AlxTopOnInterstitialAdapter.TAG)
        
        if self.interstitialAd != nil {
            NSLog("%@: adReady = YES", AlxTopOnInterstitialAdapter.TAG)
            return true
        }
        
        NSLog("%@: adReady = NO", AlxTopOnInterstitialAdapter.TAG)
        return false
    }
    
    // MARK: - Show Ad (实例方法)
    
    @objc(showInterstitialInViewController:)
    public func showInterstitial(in viewController: UIViewController) {
        NSLog("%@: showInterstitialInViewController", AlxTopOnInterstitialAdapter.TAG)
        
        DispatchQueue.main.async {
            if let interstitialAd = self.interstitialAd {
                interstitialAd.showAd(present: viewController)
                NSLog("%@: ad shown", AlxTopOnInterstitialAdapter.TAG)
            } else {
                NSLog("%@: interstitialAd is nil", AlxTopOnInterstitialAdapter.TAG)
            }
        }
    }
}
